REGRESSION(275262@main) AudioContext.start() takes longer than it used to.

jernoble opened the pull request:
#### a6d91e7d8f97f21ba3878cb1e722ae2bc75180cc
<pre>
REGRESSION(275262@main) AudioContext.start() takes longer than it used to.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272240">https://bugs.webkit.org/show_bug.cgi?id=272240</a>
<a href="https://rdar.apple.com/124071843">rdar://124071843</a>

Reviewed by Eric Carlson.

In 275262@main, SharedAudioDestination was added to reduce the runtime overhead of having
multiple AudioDestinations with the same configuration running simultaneously. The
SharedAudioDestination passes its start() and stop() completion handlers to
SharedAudioDestinationAdapter, which calls the completion handler when the underlying
AudioDestination starts and stops. But if the destination is already running, it waits
until that destination is fully &quot;configured&quot; to call the completion handler, and this
configuration happens on the audio render thread. The audio render thread spins every 3ms,
so this delay should average out to 1.5ms. However, the underlying RemoteAudioDestinationProxy
will run its completion handler without waitaing for the audio device to start, so the delay
for that class is much smaller: effecitvely the duration of the XPC call.

In the case that the AudioDestination starts and stops, continue to pass the completion
handler to that method. But in the case where the AudioDestination is already running and
the list of renderers just needs to be updated, assume that will succeed and just run the
completion handler in the next run loop.

* PerformanceTests/AudioContext/audio-context-creation.html: Added.
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestinationAdapter::callAllConfigurationHandlers): Deleted.

Canonical link: <a href="https://commits.webkit.org/277209@main">https://commits.webkit.org/277209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9ac863efc2dbcb3cf0eeb5c5e0d59acd78c2085

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46779 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25940 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49402 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49456 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49086 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30336 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23404 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38113 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47359 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22946 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40288 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19402 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41420 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4825 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51328 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21788 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18145 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45396 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23076 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44353 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10379 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23573 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Small typo in the commit message

> RemoteAudioDestinationProxy will run its completion handler without **waitaing** for the audio device to start

eric-carlson commented on PerformanceTests/AudioContext/audio-context-creation.html:
```diff
@@ -0,0 +1,46 @@
+                await context.resume();
+            }
+
+            for (let context of audioContexts) {
+                await context.suspend();
+            }
```
Style nit: braces are unnecessary

jernoble committed:
REGRESSION(275262@main) AudioContext.start() takes longer than it used to.
https://bugs.webkit.org/show_bug.cgi?id=272240
rdar://124071843

Reviewed by Eric Carlson.

In 275262@main, SharedAudioDestination was added to reduce the runtime overhead of having
multiple AudioDestinations with the same configuration running simultaneously. The
SharedAudioDestination passes its start() and stop() completion handlers to
SharedAudioDestinationAdapter, which calls the completion handler when the underlying
AudioDestination starts and stops. But if the destination is already running, it waits
until that destination is fully "configured" to call the completion handler, and this
configuration happens on the audio render thread. The audio render thread spins every 3ms,
so this delay should average out to 1.5ms. However, the underlying RemoteAudioDestinationProxy
will run its completion handler without waitaing for the audio device to start, so the delay
for that class is much smaller: effecitvely the duration of the XPC call.

In the case that the AudioDestination starts and stops, continue to pass the completion
handler to that method. But in the case where the AudioDestination is already running and
the list of renderers just needs to be updated, assume that will succeed and just run the
completion handler in the next run loop.

* PerformanceTests/AudioContext/audio-context-creation.html: Added.
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::configureRenderThread):
(WebCore::SharedAudioDestinationAdapter::render):
(WebCore::SharedAudioDestinationAdapter::callAllConfigurationHandlers): Deleted.

Canonical link: https://commits.webkit.org/277209@main
